### PR TITLE
create checkbox for items, work on updating date

### LINKED ISSUE #9: 8. As a user, I want to mark an item on my shopping list as purchased so the app can learn how often I buy different items.
AC:

* User is able to tap a checkbox or similar UI element to mark an item in the list as purchased
* Item should be shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day)

### PR DESCRIPTION
**Pull request by veeteeran**
## Description

When user taps on checkbox next to item on the list a `Date` object is created and the `lastPurchased` property in Firebase is updated. Once 24 hours have passed the boxes are unchecked.

## Related Issue

Closes #9

## Acceptance Criteria

- [x] User is able to tap a checkbox or similar UI element to mark an item in the list as purchased
- [x] Item should be shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day)

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓ | :sparkles: New feature     |

### Before
Screen Shot 2021-01-27 at 10 31 08 AM

### After

**User taps to check items**

Screen Shot 2021-01-29 at 5 09 23 PM

**After 24 hours have passed boxes are unchecked**

Screen Shot 2021-01-29 at 5 12 09 PM

## Testing Steps / QA Criteria

* One way to test the site is to click on the `details` button below to open the Netlify preview

Screen Shot 2021-01-12 at 6 05 51 PM

* Otherwise from your terminal, pull down this branch with `git pull origin DD-VT-check-item-on-list` and check that branch out with `git checkout DD-VT-check-item-on-list`
* `npm start` to launch the app.
* From add item page click item on list

### CODE REVIEW

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:46]: Instead of `onClick`, this function should be `onChange`. 👍

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:15]: Thanks for the comments! :-)

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:45]: Right now, if the user clicks on an already checked checkbox it will stay locked in the checked state. 
What if the user accidentally clicked the wrong one by mistake? Being able to clear the checkbox might be useful.

This wasn't explicitly called out in the acceptance criteria, so don't feel like you need to make this change. Just an idea worth thinking about!

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:29]: Nice use of the id field on the event to track which item to update 👍

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:19]: `today` and `daySincePurchased` are both of type `Date` right? Is it possible to remove the `.toString` and the `Date.parse` and just use them directly as Date objects? (I could be wrong, too! I'm not sure)

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:45]: Thanks @chrisdopuch. We noticed that issue as well and that was brought up during office hours today. It sounds like something that will be handled down the road

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:13]: This variable could also be a `const` since it's never reassigned.

**Review by chrisdopuch — COMMENTED** [review]
So far so good! I'll defer to the rest of the team for the initial review, just wanted to give you some early feedback. 
The demo worked great 👍 
I'll check on the 24 hour un-check tomorrow 😄

File: src/components/List.js
Comment (chrisdopuch) [reply src/components/List.js:45]: Sounds good to me, glad you are thinking about it too 🙌

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:19]: Looking into this now. I think we had that in to console log something at some point. I've deleted and it still works. Going to do some more testing before I push the change 😄

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:18]: Nit-pick: 
Any time you find yourself with a conditional or ternary, in which you are returning just `true` or `false`, you can actually just return the conditional statement itself. If you're returning false if the conditional is true, as is the case here, then you can then negate the conditional statement and return it.

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:13]: Really interesting approach here :-) I really like the idea of adding a day to the purchase day and then comparing that to today.

I think that the variable naming is a little confusing though. We're calling this variable `daysSincePurchased`, so I would expect it to be an integer indicating how many days have passed since it was purchased. However, it's actually a firebase timestamp of purchased date that is converted to a date object. Then, when we mutate it, it becomes that date plus one day.
I wonder if the team has any ideas for what would be a some good names here?

PS - one of my favorite programming quotes is: 

"There are only two hard things in Computer Science: cache invalidation and naming things."

-- Phil Karlton

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:28]: Non-rhetorical question for anyone and everyone - would it be preferable to save Date.now(), which would give us a unix timestamp, rather than `new Date()`, which appears to get converted into a Firebase timestamp (https://firebase.google.com/docs/reference/android/com/google/firebase/Timestamp)?
For me I feel like a unix timestamp would be more straightforward to deal with and leaves our logic slightly less tightly-coupled to firebase, but I don't know if there are advantages to doing it this away though that I'm not aware of.
If we were to use a Date.now(), in `checkPurchasedDate`, I believe that we would convert to a JS date with something that looks like `new Date(purchasedDate)`

File: src/components/List.js
Comment (MikeBLambert) [reply src/components/List.js:29]: I'll second that one! Nice one!

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:18]: Hi Mike. I've simplified the code a bit from the comments Chris made above. When I have this `return (today > daySincePurchased);` I'm unable to check the boxes. Not sure why this is happening. Your suggestion makes sense to me

Without a `return true;` I get this error
Screen Shot 2021-01-29 at 6 10 03 PM

Comment (MikeBLambert) [reply src/components/List.js:18]: I believe that this error indicates that you are returning undefined somehow. when you console.log(today > daySincePurchased), what does it evaluate to? Are you returning early anywhere in the function?

Comment (MikeBLambert) [reply src/components/List.js:18]: I believe that you need to switch it to `daySincePurchased > today`.

Comment (veeteeran) [reply src/components/List.js:18]: So much fun! 🤯 I need to think on this before I push any changes 😃

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:26]: @MikeBLambert dates are making my head hurt also 😅 I'm going to step away for the night and think some more. Thanks for all the  comments!

File: src/components/List.js
Comment (rachelrohrbach) [reply src/components/List.js:15]: Yes, the comments are great!

**Review by rachelrohrbach — APPROVED** [review]
This was a tough issue, so way to go! Dates are 🤯 I appreciate the comments in the code and it passes the AC!

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js]: For ideas on variable names, maybe something regarding a timestamp, like timestampOfPurchasedDate or timestampOfCurrentDate (those are a little verbose, but something in that realm or ballpark maybe) ?

**Review by LINDGRENBA — APPROVED** [review]
This was a really tough one and I know both of you put in a lot of work this week. Great job on a very difficult issue. I think there were a lot of great questions and suggestions brought up by the rest of the team about things to consider, but it works well and you've met the AC! Awesome 😄 👍

**Review by chrisdopuch — APPROVED** [review]
This is looking great! I was able to get the checkboxes to uncheck after 24 hours. Nice work y'all! 👍 
(sorry I didn't get around to the review yesterday, my night got away from me!)

File: src/components/List.js
Comment (DeeDowns) [reply src/components/List.js]: I believe we wrote it that way because we were getting confused with the logic and wanted to make it easier for our brains to understand what was going on 😅

File: src/components/List.js
Comment (DeeDowns) [reply src/components/List.js]: I agree, the naming conventions are a bit confusing, I like your suggestion Brittany

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:17]: This is so easy to read :-)

**Review by MikeBLambert — APPROVED** [review]
Looks good! Great job sticking to this one. Dates are frustrating, to say the least, and y'all did a wonderful job dealing with them! 

Note - prior to merging, you will want to resolve the merge conflicts.

**Comment by LINDGRENBA** [comment]
Just in case anyone is wondering, Viet and Dee have already fixed the error handling by switching it from a hard coded message to an empty string as initial value. Rachel and I had thought we'd taken care of that, but somehow it snuck by us. Anyways, once this branch is merged with main, that should be resolved and work as expected.

**Pull request merged**
